### PR TITLE
Shim support for acl ch.

### PR DESCRIPTION
**Pull request by rrauber**

### CODE REVIEW

File: gslib/commands/acl.py [hunk 1]
```diff
@@ -333,6 +333,20 @@ class AclCommand(Command):
       },
   )
 
+  def _get_shim_command_group(self):
+    object_or_bucket_urls = [StorageUrlFromString(i) for i in self.args]
```
Comment (NickGoog) [inline gslib/commands/acl.py:337]: `i` is usually used for index integers

File: gslib/utils/acl_helper.py [hunk 2]
```diff
@@ -187,40 +187,44 @@ def _YieldMatchingEntries(self, current_acl):
...
     elif self.scope_type == 'Project':
-      entry = entry_class(entityId=self.identifier,
-                          role=self.perm,
-                          entity=self.project_entity_prefix + self.identifier)
+      return self.project_entity_prefix + self.identifier
     elif self.scope_type == 'UserByEmail':
-      entry = entry_class(email=self.identifier,
-                          role=self.perm,
-                          entity=self.user_entity_prefix + self.identifier)
+      return self.user_entity_prefix + self.identifier
```
Comment (NickGoog) [inline gslib/utils/acl_helper.py:199]: Duplicate `return` lines. Maybe `or` statements are in order?

File: gslib/utils/acl_helper.py [hunk 3]
```diff
@@ -187,40 +187,44 @@ def _YieldMatchingEntries(self, current_acl):
...
+
+  def _AddEntry(self, current_acl, entry_class):
+    """Adds an entry to current_acl."""
+    entity = self.GetEntity()
+    entry = entry_class(role=self.perm, entity=entity)
+    if self.scope_type == 'UserById':
+      entry.entityId = self.identifier
+    elif self.scope_type == 'GroupById':
+      entry.entityId = self.identifier
+    elif self.scope_type == 'Project':
```
Comment (NickGoog) [inline gslib/utils/acl_helper.py:220]: Maybe use an `or` statement?

File: gslib/tests/test_acl.py [hunk 4]
```diff
@@ -83,7 +84,7 @@ def test_set_invalid_acl_object(self):
                             return_stderr=True,
                             expected_status=1)
     if self._use_gcloud_storage:
-      error_text = 'JSONDecodeError'
+      error_text = 'invalid JSON'
```
Comment (NickGoog) [inline gslib/tests/test_acl.py:87]: Do you want to review this? https://github.com/GoogleCloudPlatform/gsutil/pull/1722

File: gslib/tests/test_acl.py [hunk 5]
```diff
@@ -272,155 +273,179 @@ def test_acl_set_version_specific_uri(self):
...
+
+  def _MakeProjectScopeRegex(self, role, project_team, project_number):
+    template_regex = (
+        r'\{.*"entity":\s*"project-%s-%s",\s*"projectTeam":\s*\{\s*"'
+        r'projectNumber":\s*"%s",\s*"team":\s*"%s"\s*\},\s*"role":\s*"%s".*\}' %
+        (project_team, project_number, project_number, project_team, role))
+
+    return re.compile(template_regex, flags=re.DOTALL)
+
   def testAclChangeWithUserId(self):
```
Comment (NickGoog) [inline gslib/tests/test_acl.py:289]: Why was this change necessary?

File: gslib/tests/test_acl.py (hunk 4 above)
Comment (rrauber) [reply gslib/tests/test_acl.py:87]: Sure!

File: gslib/tests/test_acl.py (hunk 5 above)
Comment (rrauber) [reply gslib/tests/test_acl.py:289]: Previously these tests were basically just unit tests of the translation functionality. Now they run end to end so we can use them to confirm parity. Was there a more specific change you were referring to?

File: gslib/utils/acl_helper.py (hunk 2 above)
Comment (rrauber) [reply gslib/utils/acl_helper.py]: Sounds good!

File: gslib/commands/acl.py (hunk 1 above)
Comment (rrauber) [reply gslib/commands/acl.py]: Replaced

File: gslib/tests/test_acl.py (hunk 5 above)
Comment (NickGoog) [reply gslib/tests/test_acl.py:289]: Got it

File: gslib/tests/test_acl.py [hunk 6]
```diff
@@ -272,155 +273,179 @@ def test_acl_set_version_specific_uri(self):
...
+                               return_stdout=True)
+    self.assertNotRegex(json_text, test_regex)
 
   def testAclChangeWithGroupEmail(self):
-    change = acl_helper.AclChange(self.GROUP_TEST_ADDRESS + ':fc',
-                                  scope_type=acl_helper.ChangeType.GROUP)
-    acl = list(AclTranslation.BotoBucketAclToMessage(self.sample_uri.get_acl()))
-    change.Execute(self.sample_url, acl, 'acl', self.logger)
-    self._AssertHas(acl, 'OWNER', 'GroupByEmail', self.GROUP_TEST_ADDRESS)
+    test_regex = self._MakeScopeRegex('READER', 'group',
```
Comment (NickGoog) [inline gslib/tests/test_acl.py:351]: This changed from OWNER to READER

File: gslib/tests/test_acl.py [hunk 7]
```diff
@@ -272,155 +273,179 @@ def test_acl_set_version_specific_uri(self):
...
+    self.assertNotRegex(json_text, test_regex)
 
+  @SkipForXML('XML API does not support project scopes.')
   def testAclChangeWithProjectOwners(self):
-    change = acl_helper.AclChange(self._project_test_acl + ':READ',
-                                  scope_type=acl_helper.ChangeType.PROJECT)
-    acl = list(AclTranslation.BotoBucketAclToMessage(self.sample_uri.get_acl()))
-    change.Execute(self.sample_url, acl, 'acl', self.logger)
-    self._AssertHas(acl, 'READER', 'Project', self._project_test_acl)
+    test_regex = self._MakeProjectScopeRegex('WRITER', self._project_team,
```
[the remark this answers is not in the document]
Comment (rrauber) [reply gslib/tests/test_acl.py:394]: This was intentional, the project-owners group already has the READER role on these resources, so the first `self.assertNotRegex` call fails unless we change the role.

File: gslib/tests/test_acl.py [hunk 8]
```diff
@@ -990,3 +1022,232 @@ def test_shim_translates_acl_set_mix_buckets_and_objects_raises_error(self):
...
+                               ('GSUtil', 'hidden_shim_mode', 'dry_run')]):
+      with SetEnvironmentForTest({
+          'CLOUDSDK_CORE_PASS_CREDENTIALS_TO_GSUTIL': 'True',
+          'CLOUDSDK_ROOT_DIR': 'fake_dir',
+      }):
+        # Non-exhaustive set of strings allowed by gsutil's regex.
+        for identifier in [
+            'allauth', 'allAuthenticatedUsers', 'AllAuthenticatedUsers'
+        ]:
+          with self.subTest(identifier=identifier):
```
[the remark this answers is not in the document]
Comment (rrauber) [reply gslib/tests/test_acl.py:1172]: Didn't see any issues related to it this time around, but there were also a lot of flakes. Hopefully this run is cleaner.

**Review by NickGoog — APPROVED** [review]

File: gslib/tests/test_acl.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (rrauber) [reply gslib/tests/test_acl.py]: Removed it, there was a strange error.
